### PR TITLE
[Dataset] Add CoraFull, Amazon, KarateClub, Coauthor Dataset

### PR DESCRIPTION
**Pull request by VoVAllen**
## Description
Add CoraFull, Amazon, KarateClub, Coauthor Dataset

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
|  Datset Name   |                          Usage                           |# of graphs|Avg. # of nodes|Avg. # of edges|                 Node field                 |Edge field|Graph field|Temporal|
|----------------|----------------------------------------------------------|----------:|--------------:|--------------:|--------------------------------------------|----------|-----------|--------|
|BitcoinOTC      |BitcoinOTC()                                              |        136|        6005.00|        21210.0|                                            |h         |           |True    |
|Cora            |CoraDataset()                                             |          1|        2708.00|        10556.0|train_mask, val_mask, test_mask, label, feat|          |           |False   |
|Citeseer        |CitationGraphDataset('citeseer')                          |          1|        3327.00|         9228.0|train_mask, val_mask, test_mask, label, feat|          |           |False   |
|PubMed          |CitationGraphDataset('pubmed')                            |          1|       19717.00|        88651.0|train_mask, val_mask, test_mask, label, feat|          |           |False   |
|QM7b            |QM7b()                                                    |       7211|          15.42|          245.0|                                            |h         |label      |False   |
|Reddit          |RedditDataset()                                           |          1|      232965.00|    114615892.0|train_mask, val_mask, test_mask, feat, label|          |           |False   |
|ENZYMES         |TUDataset('ENZYMES')                                      |        600|          32.63|          124.3|feat                                        |          |label      |False   |
|DD              |TUDataset('DD')                                           |       1178|         284.32|         1431.3|feat                                        |          |label      |False   |
|COLLAB          |TUDataset('COLLAB')                                       |       5000|          74.49|         9830.0|feat                                        |          |label      |False   |
|PPI             |PPIDataset('train')/PPIDataset('valid')/PPIDataset('test')|         20|        2245.30|        63563.7|feat                                        |          |label      |False   |
|KarateClub      |KarateClub()                                              |          1|          34.00|          156.0|label                                       |          |           |False   |
|Amazon computer |AmazonCoBuy('computers')                                  |          1|       13752.00|       574418.0|feat, label                                 |          |           |False   |
|Amazon photo    |AmazonCoBuy('photo')                                      |          1|        7650.00|       287326.0|feat, label                                 |          |           |False   |
|Coauthor cs     |Coauthor('cs')                                            |          1|       18333.00|       327576.0|feat, label                                 |          |           |False   |
|Coauthor physics|Coauthor('physics')                                       |          1|       34493.00|       991848.0|feat, label                                 |          |           |False   |
|GDELT           |GDELT('train')/GDELT('valid')/GDELT('test')               |       2304|       23033.00|       811333.1|                                            |rel_type  |           |True    |
|ICEWS18         |ICEWS18('train')/ICEWS18('valid')/ICEWS18('test')         |        240|       23033.00|       192640.2|                                            |rel_type  |           |True    |
|CoraFull        |CoraFull()                                                |          1|       19793.00|       130622.0|feat, label                                 |          |           |False   |

File: examples/pytorch/diffpool/train.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def graph_classify_task(prog_args):
     perform graph classification task
     '''
 
-    dataset = tu.TUDataset(name=prog_args.dataset)
+    dataset = tu.LegacyTUDataset(name=prog_args.dataset)
```
Comment (jermainewang) [inline examples/pytorch/diffpool/train.py]: Do the `LegacyTUDataset` and `TUDataset` have the same APIs?

File: python/dgl/data/citation_graph.py [hunk 2]
```diff
@@ -29,6 +30,7 @@ def _pickle_load(pkl_file):
 
 class CitationGraphDataset(object):
```
Comment (jermainewang) [inline python/dgl/data/citation_graph.py]: Some docstring about the tensors stored in the returned DGLGraph? Example codes are also appreciated. Similar for the other datasets.

File: python/dgl/data/ppi.py [hunk 3]
```diff
@@ -136,3 +141,138 @@ def __getitem__(self, item):
             return self.valid_graphs[item], self.features[self.valid_mask_list[item]], self.valid_labels[item]
         if self.mode == 'test':
             return self.test_graphs[item], self.features[self.test_mask_list[item]], self.test_labels[item]
+
+
+class PPIDataset(object):
```
Comment (jermainewang) [inline python/dgl/data/ppi.py]: How's this different than `LegacyPPIDataset`?

Comment (VoVAllen) [reply python/dgl/data/ppi.py]: Almost same other than the `__getitem__` part. I changed it to inheritance and override `__getitem__`

File: python/dgl/data/citation_graph.py (hunk 2 above)
Comment (VoVAllen) [reply python/dgl/data/citation_graph.py]: added

File: examples/pytorch/diffpool/train.py (hunk 1 above)
Comment (VoVAllen) [reply examples/pytorch/diffpool/train.py]: The processing code are quite different. In the new `TUDataset`, I used only used raw data without adding one-hot label or constant as node features.

Comment (Mario-Liu) [reply examples/pytorch/diffpool/train.py]: hope QM9 results will be online soon

**Pull request merged**